compiler/src/iree/compiler/Codegen/Dialect/PCF/Transforms/ConvertSRefToMemRef.cpp

```diff
@@ -0,0 +1,1128 @@
+    if (solver.getExplorer().walkReturnOperands(
+            callableOp, [&](OperandRange range) {
+              int64_t currState = 0;
+              for (Value v : range) {
+                if (isa<PCF::ShapedRefType>(v.getType())) {
+                  FailureOr<MemRefType> maybeConvertedType =
+                      getConvertedType(v);
+                  if (failed(maybeConvertedType)) {
+                    return WalkResult::interrupt();
+                  }
+                  StridedLayoutState newState;
+                  newState.setAssumed(maybeConvertedType.value());
+                  newState.indicateOptimisticFixpoint();
+                  states[currState] ^= newState;
+                  ++currState;
+                }
+              }
+              return WalkResult::advance();
+            }) == TraversalResult::INCOMPLETE) {
+      return failure();
+    }
+
+    SmallVector<Type> newResultTypes(callableOp.getResultTypes());
+    int64_t currState = 0;
+    for (auto &type : newResultTypes) {
+      if (isa<PCF::ShapedRefType>(type)) {
+        auto &state = states[currState];
+        if (!state.isValidState() || !state.getAssumed()) {
+          return failure();
+        }
+        type = state.getAssumed();
+      }
+    }
+    auto newFuncType = FunctionType::get(callableOp->getContext(),
+                                         argumentTypes, newResultTypes);
+    cachedFunctionTypeMap[callableOp] = newFuncType;
+    return newFuncType;
+  }
+
+  // Returns the function type the callee of this caller should be converted to.
+  FailureOr<FunctionType>
+  lookupConvertedFunctionArgs(Operation *caller, SymbolRefAttr targetSymbol) {
+    auto callableOp =
+        solver.getExplorer()
+            .getSymbolTables()
+            .lookupNearestSymbolFrom<CallableOpInterface>(caller, targetSymbol);
+    if (!callableOp) {
+      return failure();
+    }
+    return lookupConvertedFunctionArgs(callableOp);
+  }
+
+private:
+  Explorer explorer;
+  llvm::BumpPtrAllocator allocator;
+  DFX::Solver solver;
+  llvm::SmallDenseMap<Operation *, FunctionType> cachedFunctionTypeMap;
+};
+
+//===----------------------------------------------------------------------===//
+// Conversion Patterns
+//===----------------------------------------------------------------------===//
+
+static Value castIfMismatched(OpBuilder &b, Location loc, Value v, Type t) {
+  if (v.getType() != t) {
+    assert(isa<MemRefType>(v.getType()) &&
+           "unexpected non-memref type mismatch");
+    return memref::CastOp::create(b, loc, t, v);
+  }
+  return v;
+}
+
+static void castIfMismatched(OpBuilder &b, Location loc,
+                             MutableArrayRef<Value> vals,
+                             TypeRange targetTypes) {
+  for (auto [v, t] : llvm::zip_equal(vals, targetTypes)) {
+    v = castIfMismatched(b, loc, v, t);
+  }
+}
+
+struct ConvertGenericOp : public OpConversionPattern<PCF::GenericOp> {
```

Comment:
```suggestion
struct ConvertGenericOp final : OpConversionPattern<PCF::GenericOp> {
```
also elsewhere below